httpclient5/src/main/java/org/apache/hc/client5/http/impl/ProtocolSwitchStrategy.java

```diff
@@ -45,31 +48,53 @@
 @Internal
 public final class ProtocolSwitchStrategy {
 
-    enum ProtocolSwitch { FAILURE, TLS }
+    private static final ProtocolVersionParser PARSER = ProtocolVersionParser.INSTANCE;
 
     public ProtocolVersion switchProtocol(final HttpMessage response) throws ProtocolException {
         final Iterator<String> it = MessageSupport.iterateTokens(response, HttpHeaders.UPGRADE);
-
         ProtocolVersion tlsUpgrade = null;
+        ProtocolVersion httpUpgrade = null;
+
         while (it.hasNext()) {
-            final String token = it.next();
-            if (token.startsWith("TLS")) {
-                // TODO: Improve handling of HTTP protocol token once HttpVersion has a #parse method
-                try {
-                    tlsUpgrade = token.length() == 3 ? TLS.V_1_2.getVersion() : TLS.parse(token.replace("TLS/", "TLSv"));
-                } catch (final ParseException ex) {
-                    throw new ProtocolException("Invalid protocol: " + token);
+            final String token = it.next().trim();
+            try {
+                if (token.startsWith("TLS") || token.startsWith("HTTP/")) {
```

Comment:
@arturobernalg This line looks completely unnecessary. Why? Parse first - validate next.